src/engine/cachingreader.cpp

```diff
@@ -201,23 +200,54 @@ CachingReaderChunkForOwner* CachingReader::lookupChunkAndFreshen(SINT chunkIndex
 }
 
 void CachingReader::newTrack(TrackPointer pTrack) {
+    if (m_state == State::TrackLoading) {
+        m_pNewTrack = pTrack;
+        process();
+        return;
+    }
+    m_pNewTrack.reset();
     // Feed the track to the worker as soon as possible
     // to get ready while the reader switches its internal
     // state. There are no race conditions, because the
     // reader polls the worker.
-    m_worker.newTrack(pTrack);
-    m_worker.workReady();
+    m_worker.newTrack(std::move(pTrack));
     // Don't accept any new read requests until the current
     // track has been unloaded and the new track has been
     // loaded.
     m_state = State::TrackLoading;
     // Free all chunks with sample data from the current track.
     freeAllChunks();
+    // Emit that a new track is loading, stops the current track
+    emit trackLoading();
+}
+
+void CachingReader::onTrackLoaded(
+        TrackPointer pTrack,
+        int iSampleRate,
+        int iNumSamples) {
+    // Consume all pending messages
+    process();
+    // Forward signal depending on the current state
+    if (m_state == State::TrackLoaded) {
+        emit trackLoaded(std::move(pTrack), iSampleRate, iNumSamples);
+    } else {
+        kLogger.warning()
+                << "No track loaded";
+    }
+}
+
+void CachingReader::onTrackLoadFailed(
+        TrackPointer pTrack,
+        QString reason) {
+    // Consume all pending messages
+    process();
+    // Forward signal independent of the current state
+    emit trackLoadFailed(std::move(pTrack), reason);
 }
 
 void CachingReader::process() {
     ReaderStatusUpdate update;
-    while (m_stateFIFO.read(&update, 1) == 1) {
+    while (m_readerStatusUpdateFIFO.read(&update, 1) == 1) {
         DEBUG_ASSERT(m_state != State::Idle);
```

Comment:
This assertion may occur after ejecting a track and loading a new one.  
Was this the one you have experienced?

Reply:
Added comments to the wrong/closed PR: https://github.com/mixxxdj/mixxx/pull/2305#pullrequestreview-295881182

This is the debug assertion that helped me discover the race condition. I refuse to remove it because it is correct unless someone provides a counterexample.

Reply:
I have checked the old version in 3e9e2ae2e138f75a49626a1608d3d1664ee5a079 before the fixes and I still don't see a reason why this debug assertion is not valid?